presto-orc/src/main/java/com/facebook/presto/orc/CachingStripeMetadataSource.java

```diff
@@ -106,6 +116,36 @@ public Slice getStripeFooterSlice(OrcDataSource orcDataSource, StripeId stripeId
         return inputsBuilder.build();
     }
 
+    @Override
+    public List<RowGroupIndex> getRowIndexes(
+            MetadataReader metadataReader,
+            HiveWriterVersion hiveWriterVersion,
+            StripeId stripId,
+            StreamId streamId,
+            OrcInputStream inputStream,
+            List<HiveBloomFilter> bloomFilters,
+            RuntimeStats runtimeStats)
+            throws IOException
+    {
+        if (rowGroupIndexCache.isPresent()) {
+            List<RowGroupIndex> rowGroupIndices = rowGroupIndexCache.get().getIfPresent(new StripeStreamId(stripId, streamId));
+            if (rowGroupIndices != null) {
+                runtimeStats.addMetricValue("OrcRowGroupIndexCacheHit", 1);
+                runtimeStats.addMetricValue("OrcRowGroupIndexInMemoryBytesRead", rowGroupIndices.stream().mapToLong(RowGroupIndex::getRetainedSizeInBytes).sum());
+                return rowGroupIndices;
+            }
+            else {
+                runtimeStats.addMetricValue("OrcRowGroupIndexCacheHit", 0);
+                runtimeStats.addMetricValue("OrcRowGroupIndexStorageBytesRead", inputStream.getRetainedSizeInBytes());
```

Comment:
This is incorrect metric.  OrcInputStream can be lazy and all data may not be loaded. It is better to add this metric after the rwGroupwIndices are decoded.